impl Default for ExitCode

ijackson opened the pull request:
As suggested here
  https://github.com/rust-lang/rust/pull/106425#issuecomment-1382952598

Needs FCP since this is an insta-stable impl.

Ideally we would have `impl From<ExitCode> for ExitStatus` and implement the default `ExitStatus` using that.   That is sadly not so easy because of the various strange confusions about `ExitCode` (unix: exit status) vs `ExitStatus` (unix: wait status) in the not-really-unix platforms in `library//src/sys/unix/process`.  I'll try to follow that up.

ijackson committed:
impl Default for ExitCode

As suggested here
  https://github.com/rust-lang/rust/pull/106425#issuecomment-1382952598

cuviper commented:
@rustbot label -T-libs +T-libs-api
r? libs-api

dtolnay commented:
r? dtolnay

dtolnay commented:
@rust-lang/libs-api:
@rfcbot fcp merge

Already in Rust 1.73 (current beta) we have approved `impl Default for ExitStatus` in https://github.com/rust-lang/rust/pull/106425, producing the success exit status.

Recall that [`ExitStatus`](https://doc.rust-lang.org/std/process/struct.ExitStatus.html) is the thing you get from a subprocess and [`ExitCode`](https://doc.rust-lang.org/std/process/struct.ExitCode.html) is the thing you return from `main`. :sneezing_face:

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 01494da7789e30656fd03f841943ab4df5b3b7f4 has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay 01494da7789e30656fd03f841943ab4df5b3b7f4 -->
<!-- homu: {"type":"Approved","sha":"01494da7789e30656fd03f841943ab4df5b3b7f4","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 01494da7789e30656fd03f841943ab4df5b3b7f4 with merge d96322459717767f30037d55fb3dfa79e689d04f...
<!-- homu: {"type":"BuildStarted","head_sha":"01494da7789e30656fd03f841943ab4df5b3b7f4","merge_sha":"d96322459717767f30037d55fb3dfa79e689d04f"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6509725324/job/17681999186)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6509725324/job/17681999186","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6509725324/job/17681855242) [(plain)](https://github.com/rust-lang-ci/rust/commit/d96322459717767f30037d55fb3dfa79e689d04f/checks/17681855242/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling addr2line v0.21.0
[RUSTC-TIMING] addr2line test:false 0.563
[RUSTC-TIMING] gimli test:false 5.078
[RUSTC-TIMING] object test:false 5.551
error[E0711]: feature `process-exitcode-default` is declared stable since 1.73.0-beta.1, but was previously declared stable since 1.73.0
     |
     |
1964 | #[stable(feature = "process-exitcode-default", since = "CURRENT_RUSTC_VERSION")]

For more information about this error, try `rustc --explain E0711`.
[RUSTC-TIMING] std test:false 2.003
error: could not compile `std` (lib) due to previous error
```

</details>

dtolnay reviewed (changes requested):
Looks like a conflict with #114588. They need different `feature` names because they are being stabilized in different Rust versions.

dtolnay committed:
Deduplicate std::process Default impl feature names

    error[E0711]: feature `process-exitcode-default` is declared stable since 1.74.0-beta.1, but was previously declared stable since 1.73.0
        --> library/std/src/process.rs:1964:1
         |
    1964 | #[stable(feature = "process-exitcode-default", since = "CURRENT_RUSTC_VERSION")]
         | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit b149d16d3a48672a7cfb99d19e075ff8eef4846f has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay b149d16d3a48672a7cfb99d19e075ff8eef4846f -->
<!-- homu: {"type":"Approved","sha":"b149d16d3a48672a7cfb99d19e075ff8eef4846f","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b149d16d3a48672a7cfb99d19e075ff8eef4846f with merge 58352c06497ac1cc8f7c49d30d6df967134efbaf...
<!-- homu: {"type":"BuildStarted","head_sha":"b149d16d3a48672a7cfb99d19e075ff8eef4846f","merge_sha":"58352c06497ac1cc8f7c49d30d6df967134efbaf"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6528112586/job/17725595334)
Approved by: dtolnay
Pushing 58352c06497ac1cc8f7c49d30d6df967134efbaf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6528112586/job/17725595334"},"merge_sha":"58352c06497ac1cc8f7c49d30d6df967134efbaf"} -->

rust-timer commented:
Finished benchmarking commit (58352c06497ac1cc8f7c49d30d6df967134efbaf): [comparison URL](https://perf.rust-lang.org/compare.html?start=30d310cc1f7056fa9941b9f9c415d60091ed8fec&end=58352c06497ac1cc8f7c49d30d6df967134efbaf&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=30d310cc1f7056fa9941b9f9c415d60091ed8fec&end=58352c06497ac1cc8f7c49d30d6df967134efbaf&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=30d310cc1f7056fa9941b9f9c415d60091ed8fec&end=58352c06497ac1cc8f7c49d30d6df967134efbaf&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.4% | [3.4%, 3.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.4% | [3.4%, 3.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=30d310cc1f7056fa9941b9f9c415d60091ed8fec&end=58352c06497ac1cc8f7c49d30d6df967134efbaf&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.7% | [3.7%, 3.7%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=30d310cc1f7056fa9941b9f9c415d60091ed8fec&end=58352c06497ac1cc8f7c49d30d6df967134efbaf&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 624.787s -> 624.739s (-0.01%)
**Artifact size**: 305.54 MiB -> 305.53 MiB (-0.00%)
